be/src/storage/compaction_scheduler.cpp

```diff
@@ -0,0 +1,242 @@
+        cur_alter_task->alter_state() != ALTER_FAILED) {
+        TabletManager* tablet_manager = StorageEngine::instance()->tablet_manager();
+        TabletSharedPtr related_tablet = tablet_manager->get_tablet(cur_alter_task->related_tablet_id());
+        if (related_tablet != nullptr && tablet->creation_time() > related_tablet->creation_time()) {
+            // Current tablet is newly created during schema-change or rollup, skip it
+            LOG(INFO) << "skip tablet:" << tablet->tablet_id() << " because it is newly-created for schema change";
+            return false;
+        }
+    }
+
+    std::shared_ptr<CompactionTask> compaction_task = tablet->get_compaction(false);
+    if (compaction_task) {
+        // tablet already has a running compaction task, skip it
+        LOG(INFO) << "skip tablet:" << tablet->tablet_id()
+                  << " because there is another running compaction task:" << compaction_task->task_id();
+        return false;
+    }
+    return true;
+}
+
+bool CompactionScheduler::_can_do_compaction(Tablet* tablet, bool* need_reschedule) {
+    // when the following conditions fail, should not reschedule the candidate tablet
+    *need_reschedule = false;
+    bool precondition_ok = _check_precondition(tablet);
+    if (!precondition_ok) {
+        return false;
+    }
+
+    // when the following conditions fail, should reschedule the candidate tablet
+    *need_reschedule = true;
+    DataDir* data_dir = tablet->data_dir();
+    if (data_dir->reach_capacity_limit(0)) {
+        LOG(WARNING) << "skip tablet:" << tablet->tablet_id() << " because data dir reaches capacity limit";
+        return false;
+    }
+
+    // create a new compaction task
+    std::shared_ptr<CompactionTask> compaction_task = tablet->get_compaction(true);
+
+    if (compaction_task) {
+        bool can_do = _can_do_compaction_task(tablet, compaction_task.get());
+        *need_reschedule = !can_do;
+        return can_do;
+    } else {
+        VLOG(2) << "skip tablet:" << tablet->tablet_id() << " because creating compaction task failed.";
+        return false;
+    }
+}
+
+Tablet* CompactionScheduler::_try_get_next_tablet() {
+    VLOG(2) << "try to get next qualified tablet for round:" << _round;
+    // tmp_tablets save the tmp picked candidates tablets
+    std::vector<Tablet*> tmp_tablets;
+    Tablet* tablet = nullptr;
+    bool found = false;
+    while (true) {
+        if (!_can_schedule_next()) {
+            VLOG(2) << "_can_schedule_next is false. skip";
+            break;
+        }
+        tablet = CompactionManager::instance()->pick_candidate();
+        if (!tablet) {
+            // means there no candidate tablet, break
+            break;
+        }
+        VLOG(2) << "try tablet:" << tablet->tablet_id();
+        bool need_reschedule = true;
+        found = _can_do_compaction(tablet, &need_reschedule);
+        if (need_reschedule) {
```

Comment:
check for what?

Reply:
`_can_do_compaction` has a preset that the `need_schedule` could be true only if the `found` is true. I think it may be clearer and safer to check the `found` explicitly even we have no knowledge of `_can_do_compaction`.

Reply:
when found is true, need_schedule must be false.